### PR TITLE
Symbol occurrences in file

### PR DESCRIPTION
**Pull request by tayloraswift**
adds a method `symbolOccurrences(inFilePath:)` that returns all symbol occurrences within a source file.

### CODE REVIEW

File: Sources/IndexStoreDB/IndexStoreDB.swift [hunk 1]
```diff
@@ -312,6 +312,41 @@ public final class IndexStoreDB {
     return result
   }
 
+  public func symbolProvider(for sourceFilePath: String) -> SymbolProviderKind? {
```
Comment (ahoppen) [inline Sources/IndexStoreDB/IndexStoreDB.swift:315]: I removed this function recently and it looks to me like this PR is accidentally adding it back. Could you rebase your branch to just contain your commits?

File: lib/Index/StoreSymbolRecord.cpp [hunk 2]
```diff
@@ -289,6 +289,20 @@ bool StoreSymbolRecord::foreachCoreSymbolData(function_ref<bool(StringRef USR,
   return !Err && Finished;
 }
 
+bool StoreSymbolRecord::foreachSymbolOccurrence(function_ref<bool(SymbolOccurrenceRef Occur)> Receiver) {
+  bool Finished;
+  bool Err = doForData([&](IndexRecordReader &Reader) {
+      // Return all occurrences.
```
Comment (ahoppen) [inline lib/Index/StoreSymbolRecord.cpp:295]: Indentation of the comment seems to be off

File: include/CIndexStoreDB/CIndexStoreDB.h [hunk 3]
```diff
@@ -328,6 +328,15 @@ indexstoredb_index_symbols_contained_in_file_path(_Nonnull indexstoredb_index_t
                                                   const char *_Nonnull path,
                                                   _Nonnull indexstoredb_symbol_receiver_t);
 
+/// Iterates over all the symbol occurrences contained in \p path
```
Comment (ahoppen) [inline include/CIndexStoreDB/CIndexStoreDB.h:331]: Let’s clarify that `path` is a source file path, not eg. a path to a unit file or a record file

```suggestion
/// Iterates over all the symbol occurrences contained in the source file at \p path
```

File: lib/Index/SymbolIndex.cpp [hunk 4]
```diff
@@ -500,6 +503,33 @@ bool SymbolIndexImpl::foreachSymbolInFilePath(CanonicalFilePathRef filePath,
...
+                    auto record = createVisibleProviderForCode(providerCode, reader);
+                    didFinish = record->foreachSymbolOccurrence(Receiver);
+
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    });
```
Comment (ahoppen) [inline lib/Index/SymbolIndex.cpp:528]: Could you change this to 2-space indentation?

File: Sources/IndexStoreDB/IndexStoreDB.swift (hunk 1 above)
Comment (tayloraswift) [reply Sources/IndexStoreDB/IndexStoreDB.swift]: oops, fixed

File: lib/Index/StoreSymbolRecord.cpp (hunk 2 above)
Comment (tayloraswift) [reply lib/Index/StoreSymbolRecord.cpp]: done

File: lib/Index/SymbolIndex.cpp (hunk 4 above)
Comment (tayloraswift) [reply lib/Index/SymbolIndex.cpp]: done

File: Sources/IndexStoreDB/IndexStoreDB.swift (hunk 1 above)
Comment (tayloraswift) [reply Sources/IndexStoreDB/IndexStoreDB.swift]: would squashing the commits from this PR work?

Comment (ahoppen) [reply Sources/IndexStoreDB/IndexStoreDB.swift]: Yes, squashing is encouraged.

**Review by ahoppen — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test

**Review by bnbarham — APPROVED** [review]

**Comment by tayloraswift** [comment]
thank you!!

**Pull request merged**